examples/cross-contract-low-level/src/lib.rs

```diff
@@ -42,24 +46,34 @@ impl CrossContract {
         let arr0 = arr[..pivot].to_vec();
         let arr1 = arr[pivot..].to_vec();
         let account_id = env::current_account_id();
-        let prepaid_gas = env::prepaid_gas();
+        let gas_to_pass = Gas(match pivot {
+            1 => 1 * TGAS,
+            2 => 40 * TGAS,
+            // TODO: make work with input arrays of length 5, maybe 6
```

Comment:
Why changing to static gas for these? Will it not work if you make it proportional to the amount of prepaid gas (like it was before) and just passing in more gas for each call? Or does it hit some other limit?

Reply:
It took some work to get this contract to work with real gas values. If you look at the old simulation tests, you'll see that they only passed because the set gas price to 0 and max attached gas to ~10<sup>19</sup>. 

While figuring out realistic gas values, it became clear that the last steps in the recursion, arrays of length 1, require barely any gas. We were reserving too much gas for this step, using the simplistic previous formula, which cause earlier steps in the recursion to easily run into gas limits. Walking through a concrete example:

* With array of length 4 and 160Tgas, 
* you end up with 40Tgas passed to the depth-2 calls to `merge_sort`,
* which then try to pass 10Tgas each to the last calls to `merge_sort` and `merge`, when these only needed 1Tgas, 
* and just queueing all those cross-contract calls in the depth-2 calls to `merge_sort` requires 29Tgas, 
* so the whole thing fails

You can [see these gas values in Explorer](https://explorer.testnet.near.org/transactions/431CmS7pXjPtbvHD8shPSS5QH59rQhsV9LKsqzGP1V7S).

If we had a third number here, we might be able to determine a better formula. But with only two numbers in the sequence, it's hard to predict what the formula should be.

Reply:
@austinabell FYI. I don't think we should be modifying merge sort example like that for several reasons:
* An example that only works for array of length 4 is not really a good example. It undermines the whole point that shows that NEAR can work with deeply nested and branching function calls. We use this example in the workshops to impress people, e.g. when we give them assignment to write MapReduce on NEAR;
* This new added logic makes the code more complex which is not good, since the examples are supposed to be as simple as possible. Besides, I am not sure the following does anything if we already limit array size to 4:
```rust
1 => TGAS * 1,
2 => TGAS * 40,
```

Base function call cost is ~5Tgas. Callback cost is ~10Tgas (once @Longarithm 's protocol change gets into Mainnet it will be negligible).

So we should be able to do arrays of size ~16 even with the current costs, if Wasm operation cost is negligible (5+20+20*2+20*4+20*8 ~=300Tgas). We can significantly reduce the cost of Wasm execution itself by using Borsh serialization for all arguments and return types, and if we need to have JSON interface to call the contract from near-cli then we can separate `merge_sort` into `merge_sort` and `internal_merge_sort`.